File: server/r1cord_server/pipeline/writers.py
"""Claude Code / Codex / Grok Build adapters and summary.md validation."""

from __future__ import annotations

import os
import re
import shutil
import subprocess
import sys
from collections.abc import Callable
from pathlib import Path
from typing import Protocol

PROMPT = "Read INSTRUCTIONS.md in this folder and do exactly what it says."

_IMAGE_RE = re.compile(r"!\[([^\]]*)\]\(([^)]+)\)")


class WriterError(Exception):
    """Writer CLI failed, timed out, or summary.md is missing/invalid."""


class WriterConfig(Protocol):
    claude_cmd: str
    codex_cmd: str
    grok_cmd: str


def _resolve_cmd(cmd: str) -> str:
    found = shutil.which(cmd)
    if found:
        return found
    if sys.platform == "win32":
        stem, ext = os.path.splitext(cmd)
        if not ext:
            for shim in (".cmd", ".exe", ".bat"):
                found = shutil.which(cmd + shim)
                if found:
                    return found
        elif ext.lower() not in {".cmd", ".exe", ".bat"}:
            for shim in (".cmd", ".exe"):
                found = shutil.which(stem + shim)
                if found:
                    return found
    raise WriterError(f"writer binary not found: {cmd}")


# Console CLIs spawned from a console-less pythonw server would each open a console window.
_NO_WINDOW = getattr(subprocess, "CREATE_NO_WINDOW", 0)


def _kill_process_tree(proc: subprocess.Popen) -> None:
    pid = proc.pid
    if sys.platform == "win32":
        subprocess.run(
            ["taskkill", "/T", "/F", "/PID", str(pid)],
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            check=False,
            creationflags=_NO_WINDOW,
        )
    else:
        proc.kill()
    try:
        proc.wait(timeout=10)
    except subprocess.TimeoutExpired:
        proc.kill()


def _tail_writer_log(work_dir: Path, n: int = 20) -> str:
    path = work_dir / "writer.log"
    if not path.is_file():
        return ""
    lines = path.read_text(encoding="utf-8", errors="replace").splitlines()
    return "\n".join(lines[-n:])


def _image_target_path(work_dir: Path, target: str) -> Path | None:
    t = target.strip()
    if t.startswith("<") and t.endswith(">"):
        t = t[1:-1].strip()
    if not t:
        return None
    # Drop an optional title after the path.
    if t[0] in {'"', "'"}:
        quote = t[0]
        end = t.find(quote, 1)
        t = t[1:end] if end != -1 else t[1:]
    else:
        t = t.split()[0]
    t = t.replace("\\", "/")
    if t.startswith("./"):
        t = t[2:]
    photos = (work_dir / "photos").resolve()
    candidate = (work_dir / t).resolve()
    try:
        candidate.relative_to(photos)
    except ValueError:
        return None
    return candidate


def validate_summary(work_dir: Path, log: Callable[[str], None]) -> Path:
    """Require a non-empty summary.md; rewrite broken images; ensure [brand-header]."""
    work_dir = Path(work_dir)
    path = work_dir / "summary.md"
    if not path.is_file():
        tail = _tail_writer_log(work_dir)
        extra = f"\n{tail}" if tail else ""
        raise WriterError(f"summary.md missing at {path}{extra}")
    raw = path.read_text(encoding="utf-8")
    if not raw.strip():
        tail = _tail_writer_log(work_dir)
        extra = f"\n{tail}" if tail else ""
        raise WriterError(f"summary.md is empty at {path}{extra}")

    def _replace(match: re.Match[str]) -> str:
        alt, target = match.group(1), match.group(2)
        resolved = _image_target_path(work_dir, target)
        if resolved is not None and resolved.is_file():
            return match.group(0)
        log(f"broken image link rewritten: ![{alt}]({target}) -> *{alt}*")
        return f"*{alt}*"

    new = _IMAGE_RE.sub(_replace, raw)
    lines = new.splitlines()
    first_nonempty = next((ln.strip() for ln in lines if ln.strip()), "")
    if first_nonempty != "[brand-header]":
        log("inserted [brand-header] as first line")
        new = "[brand-header]\n" + new
        if not new.endswith("\n") and raw.endswith("\n"):
            new += "\n"
    elif not new.endswith("\n") and raw.endswith("\n"):
        new += "\n"
    if new != raw:
        path.write_text(new, encoding="utf-8")
    return path


def _argv_for(writer: str, work_dir: Path, config: WriterConfig) -> list[str]:
    if writer == "claude_code":
        bin_path = _resolve_cmd(config.claude_cmd)
        return [
            bin_path,
            "-p",
            PROMPT,
            "--add-dir",
            str(work_dir),
            "--allowedTools",
            "Read,Write,Edit,Glob,Grep",
            "--permission-mode",
            "acceptEdits",
        ]
    if writer == "codex":
        bin_path = _resolve_cmd(config.codex_cmd)
        return [
            bin_path,
            "exec",
            "-C",
            str(work_dir),
            "--sandbox",
            "workspace-write",
            "--skip-git-repo-check",
            PROMPT,
        ]
    if writer == "grok_build":
        bin_path = _resolve_cmd(config.grok_cmd)
        prompt_path = work_dir / "prompt.txt"
        prompt_path.write_text(PROMPT + "\n", encoding="utf-8")
        work_fwd = work_dir.resolve().as_posix()
        prompt_fwd = prompt_path.resolve().as_posix()
        return [
            bin_path,
            "--prompt-file",
            prompt_fwd,
            "--cwd",
            work_fwd,
            "--permission-mode",
            "bypassPermissions",
            "--disable-web-search",
            "--no-subagents",
            "--max-turns",
            "40",
        ]
    raise WriterError(f"unknown writer: {writer!r} (expected claude_code|codex|grok_build)")


def summarize(
    work_dir: Path,
    *,
    writer: str,
    timeout_s: int,
    config: WriterConfig,
    log: Callable[[str], None],
) -> Path:
    """Run the writer CLI in `work_dir` and return the validated summary.md path."""
    work_dir = Path(work_dir)
    work_dir.mkdir(parents=True, exist_ok=True)
    argv = _argv_for(writer, work_dir, config)
    log_path = work_dir / "writer.log"
    argv_line = "argv: " + repr(argv)
    log(argv_line)
    with log_path.open("a", encoding="utf-8", errors="replace") as fh:
        fh.write(argv_line + "\n")
        fh.flush()
        proc = subprocess.Popen(  # noqa: S603 — argv list, shell=False
            argv,
            cwd=str(work_dir),
            stdin=subprocess.DEVNULL,
            stdout=fh,
            stderr=subprocess.STDOUT,
            shell=False,
            creationflags=_NO_WINDOW,
        )
        try:
            rc = proc.wait(timeout=timeout_s)
        except subprocess.TimeoutExpired as exc:
            _kill_process_tree(proc)
            tail = _tail_writer_log(work_dir)
            extra = f"\n{tail}" if tail else ""
            raise WriterError(f"writer {writer} timed out after {timeout_s}s{extra}") from exc
    if rc != 0:
        tail = _tail_writer_log(work_dir)
        extra = f"\n{tail}" if tail else ""
        raise WriterError(f"writer {writer} exited {rc}{extra}")
    return validate_summary(work_dir, log)

File: desktop/src/core/pipeline/writers.js
// Port of r1cord_server/pipeline/writers.py: Claude Code / Codex / Grok Build adapters, `<kind>.md`
// validation, and the transcript page.

const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');
const cli = require('../cli');
const { PY_SPACE, isFile, pySplit, pySplitlines, pyStrRepr, pyStrip, readText, platformText, writeText } = require('./compat');

const PROMPT = 'Read INSTRUCTIONS.md in this folder and do exactly what it says.';

const IMAGE_RE = /!\[([^\]]*)\]\(([^)]+)\)/g;

// Writer CLI failed, timed out, or `<kind>.md` is missing/invalid.
class WriterError extends Error {
  constructor(message) {
    super(message);
    this.name = 'WriterError';
  }
}

/**
 * The writer binary for a configured command: shutil.which, then on Windows the .cmd/.exe/.bat
 * shims npm and friends install.
 * @param {string} cmd
 * @param {{platform?: string, which?: Function}} [options]
 * @returns {string}
 */
function resolveCmd(cmd, { platform = process.platform, which = cli.which } = {}) {
  let found = which(cmd);
  if (found) return found;
  if (platform === 'win32') {
    const ext = path.win32.extname(cmd);
    const stem = ext ? cmd.slice(0, -ext.length) : cmd;
    const shims = !ext ? ['.cmd', '.exe', '.bat'] : ['.cmd', '.exe', '.bat'].includes(ext.toLowerCase()) ? [] : ['.cmd', '.exe'];
    for (const shim of shims) {
      found = which((ext ? stem : cmd) + shim);
      if (found) return found;
    }
  }
  throw new WriterError(`writer binary not found: ${cmd}`);
}

function tailWriterLog(workDir, n = 20) {
  const file = path.join(workDir, 'writer.log');
  if (!isFile(file)) return '';
  const lines = pySplitlines(readText(file, { errors: 'replace' }));
  return lines.slice(-n).join('\n');
}

function withTail(message, workDir) {
  const tail = tailWriterLog(workDir);
  return tail ? `${message}\n${tail}` : message;
}

// Path.resolve(): the real path of the longest existing prefix, lexical for the rest.
function resolvePath(target) {
  const absolute = path.resolve(target);
  let existing = absolute;
  const rest = [];
  for (;;) {
    try {
      return path.join(fs.realpathSync.native(existing), ...rest);
    } catch (_error) {
      const parent = path.dirname(existing);
      if (parent === existing) return absolute;
      rest.unshift(path.basename(existing));
      existing = parent;
    }
  }
}

function isInside(root, candidate) {
  const norm = (p) => (process.platform === 'win32' ? p.toLowerCase() : p);
  const rel = path.relative(norm(root), norm(candidate));
  return rel === '' || (!rel.startsWith('..') && !path.isAbsolute(rel));
}

function imageTargetPath(workDir, target) {
  let t = pyStrip(target);
  if (t.startsWith('<') && t.endsWith('>')) t = pyStrip(t.slice(1, -1));
  if (!t) return null;
  // Drop an optional title after the path.
  if (t[0] === '"' || t[0] === "'") {
    const end = t.indexOf(t[0], 1);
    t = end !== -1 ? t.slice(1, end) : t.slice(1);
  } else {
    t = pySplit(t)[0];
  }
  t = t.replace(/\\/g, '/');
  if (t.startsWith('./')) t = t.slice(2);
  const photos = resolvePath(path.join(workDir, 'photos'));
  const candidate = resolvePath(path.resolve(workDir, t));
  return isInside(photos, candidate) ? candidate : null;
}

/**
 * Require a non-empty `<kind>.md`; rewrite image links that do not resolve to a listed photo.
 * @param {string} workDir
 * @param {string} kind
 * @param {(line: string) => void} log
 * @returns {string} the review's path
 */
function validateReview(workDir, kind, log) {
  const name = `${kind}.md`;
  const file = path.join(workDir, name);
  if (!isFile(file)) throw new WriterError(withTail(`${name} missing at ${file}`, workDir));
  const raw = readText(file);
  if (!pyStrip(raw)) throw new WriterError(withTail(`${name} is empty at ${file}`, workDir));

  const rewritten = raw.replace(IMAGE_RE, (whole, alt, target) => {
    const resolved = imageTargetPath(workDir, target);
    if (resolved !== null && isFile(resolved)) return whole;
    log(`broken image link rewritten: ![${alt}](${target}) -> *${alt}*`);
    return `*${alt}*`;
  });
  if (rewritten !== raw) writeText(file, rewritten);
  return file;
}

// Path.resolve().as_posix() for grok, whose parser strips backslashes.
function posixResolved(target) {
  return resolvePath(target).replace(/\\/g, '/');
}

/**
 * The argv for one writer run in `workDir`.
 * @param {string} writer claude_code | codex | grok_build
 * @param {string} workDir
 * @param {{claude_cmd: string, codex_cmd: string, grok_cmd: string}} config
 * @param {{resolveCmd?: Function}} [options]
 * @returns {string[]}
 */
function argvFor(writer, workDir, config, { resolveCmd: resolve = resolveCmd } = {}) {
  if (writer === 'claude_code') {
    return [
      resolve(config.claude_cmd),
      '-p',
      PROMPT,
      '--add-dir',
      workDir,
      '--allowedTools',
      'Read,Write,Edit,Glob,Grep',
      '--permission-mode',
      'acceptEdits',
    ];
  }
  if (writer === 'codex') {
    return [resolve(config.codex_cmd), 'exec', '-C', workDir, '--sandbox', 'workspace-write', '--skip-git-repo-check', PROMPT];
  }
  if (writer === 'grok_build') {
    const bin = resolve(config.grok_cmd);
    const promptFile = path.join(workDir, 'prompt.txt');
    writeText(promptFile, `${PROMPT}\n`);
    return [
      bin,
      '--prompt-file',
      posixResolved(promptFile),
      '--cwd',
      posixResolved(workDir),
      '--permission-mode',
      'bypassPermissions',
      '--disable-web-search',
      '--no-subagents',
      '--max-turns',
      '40',
    ];
  }
  throw new WriterError(`unknown writer: ${pyStrRepr(writer)} (expected claude_code|codex|grok_build)`);
}

function exitCode(code, signal) {
  if (code !== null) return code;
  const number = signal ? os.constants.signals[signal] : undefined;
  return number ? -number : -1;
}

/**
 * Run the writer CLI in `workDir` (holding that review's INSTRUCTIONS.md) and return the validated
 * `<kind>.md` path. Output goes to `writer.log`; a timeout or abort kills the whole process tree.
 * @param {string} workDir
 * @param {{kind: string, writer: string, timeoutS: number, config: object, log: (line: string) => void,
 *   signal?: AbortSignal, argvFor?: Function}} options
 * @returns {Promise<string>}
 */
async function runWriter(workDir, { kind, writer, timeoutS, config, log, signal = null, argvFor: makeArgv = argvFor }) {
  fs.mkdirSync(workDir, { recursive: true });
  const argv = makeArgv(writer, workDir, config);
  const logPath = path.join(workDir, 'writer.log');
  const argvLine = `argv: [${argv.map(pyStrRepr).join(', ')}]`;
  log(argvLine);
  const fd = fs.openSync(logPath, 'a');
  let outcome;
  try {
    fs.writeSync(fd, platformText(`${argvLine}\n`));
    outcome = await waitForChild(cli.spawn(argv, { cwd: workDir, stdio: ['ignore', fd, fd] }), timeoutS, signal);
  } finally {
    fs.closeSync(fd);
  }
  if (outcome.aborted) throw signal.reason instanceof Error ? signal.reason : new WriterError(`writer ${writer} was stopped`);
  if (outcome.timedOut) throw new WriterError(withTail(`writer ${writer} timed out after ${timeoutS}s`, workDir));
  if (outcome.rc !== 0) throw new WriterError(withTail(`writer ${writer} exited ${outcome.rc}`, workDir));
  return validateReview(workDir, kind, log);
}

// Wait for the child; on timeout or abort kill its tree and wait up to 10 s for it to go.
function waitForChild(child, timeoutS, signal) {
  return new Promise((resolve, reject) => {
    let timedOut = false;
    let aborted = false;
    let exited = false;
    let killTimer;
    const kill = () => {
      cli.killTree(child).then(() => {
        if (!exited) killTimer = setTimeout(() => child.kill('SIGKILL'), 10_000).unref();
      });
    };
    const timer = setTimeout(() => {
      timedOut = true;
      kill();
    }, timeoutS * 1000);
    const onAbort = () => {
      aborted = true;
      kill();
    };
    if (signal) {
      if (signal.aborted) onAbort();
      else signal.addEventListener('abort', onAbort, { once: true });
    }
    const done = () => {
      exited = true;
      clearTimeout(timer);
      clearTimeout(killTimer);
      if (signal) signal.removeEventListener('abort', onAbort);
    };
    child.once('error', (error) => {
      done();
      reject(error);
    });
    child.once('exit', (code, sig) => {
      done();
      resolve({ rc: exitCode(code, sig), timedOut, aborted });
    });
  });
}

const MD_SPECIAL = /([\\`*_[\]<>#|])/g;
// `^` of a MULTILINE Python pattern: the start of the text or right after '\n' (not '\r').
const MD_LINE_START = new RegExp(`(?<![^\\n])([${PY_SPACE}]*)(?:([-+])|(\\p{Nd}+)([.)]))(?=[${PY_SPACE}]|$)`, 'gu');

// Spoken words, not Markdown: keep a transcript line from turning into a heading, list, quote,
// link or tag when the page is rendered.
function escapeMarkdown(text) {
  const special = text.replace(MD_SPECIAL, '\\$1');
  return special.replace(MD_LINE_START, (_whole, lead, bullet, digits, mark) => (bullet ? `${lead}\\${bullet}` : `${lead}${digits}\\${mark}`));
}

/**
 * The transcript page source: `# title`, a date · duration line, then the transcript's paragraphs.
 * @param {string} title
 * @param {string} when
 * @param {string} transcript
 * @returns {string}
 */
function transcriptMarkdown(title, when, transcript) {
  const paragraphs = transcript
    .replace(/\r\n/g, '\n')
    .split('\n\n')
    .map(pyStrip)
    .filter(Boolean);
  const body = paragraphs.map(escapeMarkdown).join('\n\n');
  const parts = [`# ${title}`, ''];
  if (when) parts.push(`*${when}*`, '');
  parts.push(body || '*No speech was recognized.*');
  return `${parts.join('\n')}\n`;
}

module.exports = {
  PROMPT,
  WriterError,
  argvFor,
  resolveCmd,
  runWriter,
  transcriptMarkdown,
  validateReview,
};
